File: sdk/node/examples/test_parallel_processing.ts
#!/usr/bin/env npx tsx
/**
 * A3S Code Node.js SDK - Parallel Task Processing Integration Test (TypeScript)
 *
 * Demonstrates parallel processing of multiple tasks using A3S Lane queue.
 * Tests concurrent file analysis, code review, and documentation generation.
 *
 * Run with: npx tsx examples/test_parallel_processing.ts
 */

import { Agent, Session, AgentResult } from '../index.js';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

interface TaskResult {
  taskNum: number;
  success: boolean;
  result?: AgentResult;
  error?: Error;
}

interface PriorityTaskResult {
  taskNum: number;
  type: string;
  success: boolean;
  result?: AgentResult;
  error?: Error;
}

interface QueueStats {
  totalProcessed: number;
  totalFailed: number;
  dlqSize: number;
}

class ParallelProcessingTest {
  private readonly agent: Agent;
  private readonly configPath: string;

  constructor(agent: Agent, configPath: string) {
    this.agent = agent;
    this.configPath = configPath;
  }

  /**
   * Find config file in home directory or project root.
   */
  static findConfigPath(): string {
    const homeConfig: string = path.join(os.homedir(), '.a3s', 'config.hcl');
    if (fs.existsSync(homeConfig)) {
      return homeConfig;
    }

    // Try project root (6 levels up)
    const projectConfig: string = path.join(__dirname, '..', '..', '..', '..', '..', '..', '.a3s', 'config.hcl');
    if (fs.existsSync(projectConfig)) {
      return projectConfig;
    }

    throw new Error('Config file not found. Please create ~/.a3s/config.hcl');
  }

  /**
   * Test 1: Sequential processing (baseline).
   */
  async testSequentialProcessing(): Promise<void> {
    console.log('\n📦 Test 1: Sequential Processing (Baseline)');
    console.log('-'.repeat(80));

    const session: Session = this.agent.session('.');

    const tasks: string[] = [
      'Count the number of JavaScript files in this project',
      'Find all TODO comments in JavaScript files',
      'List all exported functions in the main module',
    ];

    const start: number = Date.now();
    console.log(`Processing ${tasks.length} tasks sequentially...`);

    for (let i = 0; i < tasks.length; i++) {
      console.log(`  Task ${i + 1}: ${tasks[i]}`);
      const result: AgentResult = await session.send(tasks[i]);
      console.log(`  ✓ Completed: ${result.text.length} chars`);
    }

    const duration: number = (Date.now() - start) / 1000;
    console.log(`\n✓ Sequential processing took: ${duration.toFixed(2)}s`);
    console.log('\n✅ Test 1 passed: Sequential processing works\n');
  }

  /**
   * Test 2: Parallel processing with queue.
   */
  async testParallelProcessing(): Promise<void> {
    console.log('⚡ Test 2: Parallel Processing with Queue');
    console.log('-'.repeat(80));

    // Configure queue for parallel processing
    const session: Session = this.agent.session('.', {
      queueConfig: {
        queryConcurrency: 3,      // Allow 3 concurrent query operations
        executeConcurrency: 2,    // Allow 2 concurrent execute operations
        enableMetrics: true,      // Enable metrics collection
        enableDlq: true,          // Enable dead letter queue
      }
    });

    const tasks: string[] = [
      'Count the number of JavaScript files in this project',
      'Find all TODO comments in JavaScript files',
      'List all exported functions in the main module',
      'Find all async functions in the codebase',
      'Count lines of code in all JavaScript files',
    ];

    const start: number = Date.now();
    console.log(`Processing ${tasks.length} tasks in parallel...`);

    // Queue all tasks
    for (let i = 0; i < tasks.length; i++) {
      console.log(`  Queuing task ${i + 1}: ${tasks[i]}`);
    }

    console.log('\n  Waiting for all tasks to complete...');

    // Process tasks concurrently
    const promises: Promise<TaskResult>[] = tasks.map((task: string, i: number): Promise<TaskResult> =>
      session.send(task)
        .then((result: AgentResult): TaskResult => ({ taskNum: i + 1, success: true, result }))
        .catch((error: Error): TaskResult => ({ taskNum: i + 1, success: false, error }))
    );

    const results: TaskResult[] = await Promise.all(promises);

    for (const { taskNum, success, result, error } of results) {
      if (success) {
        console.log(`  ✓ Task ${taskNum} completed: ${result!.text.length} chars`);
      } else {
        console.log(`  ✗ Task ${taskNum} failed: ${error!.message}`);
      }
    }

    const duration: number = (Date.now() - start) / 1000;
    console.log(`\n✓ Parallel processing took: ${duration.toFixed(2)}s`);

    // Check queue stats
    if (session.hasQueue && session.hasQueue()) {
      const stats = session.queueStats() as unknown as QueueStats;
      console.log('\n📊 Queue Statistics:');
      console.log(`  Total processed: ${stats.totalProcessed}`);
      console.log(`  Total failed: ${stats.totalFailed}`);
      console.log(`  DLQ size: ${stats.dlqSize}`);
    }

    console.log('\n✅ Test 2 passed: Parallel processing with queue works\n');
  }

  /**
   * Test 3: Parallel processing with priority lanes.
   */
  async testPriorityLanes(): Promise<void> {
    console.log('🎯 Test 3: Parallel Processing with Priority Lanes');
    console.log('-'.repeat(80));

    const session: Session = this.agent.session('.', {
      queueConfig: {
        controlConcurrency: 1,    // P0: Control operations
        queryConcurrency: 3,      // P1: Query operations (highest concurrency)
        executeConcurrency: 2,    // P2: Execute operations
        generateConcurrency: 1,   // P3: Generate operations
        enableMetrics: true,
        enableDlq: true,
      }
    });

    console.log('Testing priority-based task execution...');
    console.log('  P1 (Query): 3 concurrent tasks');
    console.log('  P2 (Execute): 2 concurrent tasks');
    console.log('  P3 (Generate): 1 concurrent task');

    const start: number = Date.now();

    // Mix of different task types
    const tasks: Array<{ type: string; text: string }> = [
      { type: 'Query', text: 'How many JavaScript files are there?' },
      { type: 'Query', text: 'What is the project structure?' },
      { type: 'Query', text: 'List all dependencies' },
      { type: 'Execute', text: 'Create a summary of the codebase' },
      { type: 'Execute', text: 'Analyze code complexity' },
    ];

    for (let i = 0; i < tasks.length; i++) {
      console.log(`  Queuing ${tasks[i].type} task ${i + 1}: ${tasks[i].text}`);
    }

    console.log('\n  Waiting for all tasks to complete...');

    const promises: Promise<PriorityTaskResult>[] = tasks.map((task: { type: string; text: string }, i: number): Promise<PriorityTaskResult> =>
      session.send(task.text)
        .then((result: AgentResult): PriorityTaskResult => ({ taskNum: i + 1, type: task.type, success: true, result }))
        .catch((error: Error): PriorityTaskResult => ({ taskNum: i + 1, type: task.type, success: false, error }))
    );

    const results: PriorityTaskResult[] = await Promise.all(promises);

    for (const { taskNum, type, success, result, error } of results) {
      if (success) {
        console.log(`  ✓ Task ${taskNum} (${type}) completed: ${result!.text.length} chars`);
      } else {
        console.log(`  ✗ Task ${taskNum} (${type}) failed: ${error!.message}`);
      }
    }

    const duration: number = (Date.now() - start) / 1000;
    console.log(`\n✓ Priority-based processing took: ${duration.toFixed(2)}s`);
    console.log('\n✅ Test 3 passed: Priority lanes work correctly\n');
  }

  /**
   * Test 4: Parallel processing with retry policy.
   */
  async testRetryPolicy(): Promise<void> {
    console.log('🔄 Test 4: Parallel Processing with Retry Policy');
    console.log('-'.repeat(80));

    const session: Session = this.agent.session('.', {
      queueConfig: {
        queryConcurrency: 3,
        enableMetrics: true,
        enableDlq: true,
      }
    });

    console.log('Testing retry policy with exponential backoff...');
    console.log('  Note: Retry policy configured at queue level');
    console.log('  Max retries: 3 (default)');
    console.log('  Strategy: exponential (default)');

    const tasks: string[] = [
      'Analyze the main function',
      'Find all error types',
      'List all test functions',
    ];

    const start: number = Date.now();

    for (let i = 0; i < tasks.length; i++) {
      console.log(`  Queuing task ${i + 1}: ${tasks[i]}`);
    }

    console.log('\n  Waiting for all tasks to complete...');

    const promises: Promise<TaskResult>[] = tasks.map((task: string, i: number): Promise<TaskResult> =>
      session.send(task)
        .then((result: AgentResult): TaskResult => ({ taskNum: i + 1, success: true, result }))
        .catch((error: Error): TaskResult => ({ taskNum: i + 1, success: false, error }))
    );

    const results: TaskResult[] = await Promise.all(promises);

    for (const { taskNum, success, result, error } of results) {
      if (success) {
        console.log(`  ✓ Task ${taskNum} completed: ${result!.text.length} chars`);
      } else {
        console.log(`  ✗ Task ${taskNum} failed: ${error!.message}`);
      }
    }

    const duration: number = (Date.now() - start) / 1000;
    console.log(`\n✓ Processing with retry took: ${duration.toFixed(2)}s`);

    if (session.hasQueue && session.hasQueue()) {
      const stats = session.queueStats() as unknown as QueueStats;
      console.log('\n📊 Final Queue Statistics:');
      console.log(`  Total processed: ${stats.totalProcessed}`);
      console.log(`  Total failed: ${stats.totalFailed}`);
      console.log(`  DLQ size: ${stats.dlqSize}`);
    }

    console.log('\n✅ Test 4 passed: Retry policy works correctly\n');
  }

  /**
   * Run all tests.
   */
  async runAll(): Promise<void> {
    console.log('🚀 A3S Code Node.js SDK - Parallel Task Processing Integration Test\n');
    console.log('='.repeat(80));
    console.log(`📄 Using config: ${this.configPath}`);
    console.log('='.repeat(80));

    await this.testSequentialProcessing();
    await this.testParallelProcessing();
    await this.testPriorityLanes();
    await this.testRetryPolicy();

    console.log();
    console.log('='.repeat(80));
    console.log('✅ All parallel processing tests completed successfully!');
    console.log('='.repeat(80));
  }
}

/**
 * Main test runner.
 */
async function main(): Promise<void> {
  const configPath: string = ParallelProcessingTest.findConfigPath();
  const agent: Agent = await Agent.create(configPath);
  const test = new ParallelProcessingTest(agent, configPath);
  await test.runAll();
}

// Run tests
main().catch((err: unknown) => {
  console.error('❌ Test failed:', err);
  process.exit(1);
});

File: sdk/python/examples/test_parallel_processing.py
#!/usr/bin/env python3
"""
A3S Code Python SDK - Parallel Task Processing Integration Test

Demonstrates parallel processing of multiple tasks using A3S Lane queue.
Tests concurrent file analysis, code review, and documentation generation.

Run with: python examples/test_parallel_processing.py
"""

import asyncio
import time
from pathlib import Path
from a3s_code import Agent, SessionQueueConfig


class ParallelProcessingTest:
    """Integration test for parallel task processing with the A3S Lane queue."""

    def __init__(self, agent: Agent, config_path: str) -> None:
        self.agent = agent
        self.config_path = config_path

    @staticmethod
    def find_config_path() -> str:
        """Find config file in home directory or project root."""
        home_config = Path.home() / ".a3s" / "config.hcl"
        if home_config.exists():
            return str(home_config)

        # Try project root (6 levels up)
        project_config = Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "config.hcl"
        if project_config.exists():
            return str(project_config)

        raise FileNotFoundError("Config file not found. Please create ~/.a3s/config.hcl")

    async def test_sequential_processing(self) -> None:
        """Test 1: Sequential processing (baseline)."""
        print("\n📦 Test 1: Sequential Processing (Baseline)")
        print("-" * 80)

        session = self.agent.session(".")

        tasks = [
            "Count the number of Python files in this project",
            "Find all TODO comments in Python files",
            "List all classes in the main module",
        ]

        start = time.time()
        print(f"Processing {len(tasks)} tasks sequentially...")

        for i, task in enumerate(tasks, 1):
            print(f"  Task {i}: {task}")
            result = session.send(task)
            print(f"  ✓ Completed: {len(result.text)} chars")

        duration = time.time() - start
        print(f"\n✓ Sequential processing took: {duration:.2f}s")
        print("\n✅ Test 1 passed: Sequential processing works\n")

    async def test_parallel_processing(self) -> None:
        """Test 2: Parallel processing with queue."""
        print("⚡ Test 2: Parallel Processing with Queue")
        print("-" * 80)

        # Configure queue for parallel processing
        queue_config = SessionQueueConfig()
        queue_config.set_query_concurrency(3)      # Allow 3 concurrent query operations
        queue_config.set_execute_concurrency(2)    # Allow 2 concurrent execute operations
        queue_config.enable_metrics()              # Enable metrics collection
        queue_config.enable_dlq()                  # Enable dead letter queue

        session = self.agent.session(".", queue_config=queue_config)

        tasks = [
            "Count the number of Python files in this project",
            "Find all TODO comments in Python files",
            "List all classes in the main module",
            "Find all async functions in the codebase",
            "Count lines of code in all Python files",
        ]

        start = time.time()
        print(f"Processing {len(tasks)} tasks in parallel...")

        # Create async tasks
        async def process_task(task_num: int, task_text: str) -> tuple:
            """Process a single task."""
            result = session.send(task_text)
            return task_num, result

        # Queue all tasks
        for i, task in enumerate(tasks, 1):
            print(f"  Queuing task {i}: {task}")

        print("\n  Waiting for all tasks to complete...")

        # Process tasks concurrently using asyncio
        async_tasks = [
            asyncio.create_task(asyncio.to_thread(process_task, i, task))
            for i, task in enumerate(tasks, 1)
        ]

        results = await asyncio.gather(*async_tasks, return_exceptions=True)

        for task_num, result in results:
            if isinstance(result, Exception):
                print(f"  ✗ Task {task_num} failed: {result}")
            else:
                print(f"  ✓ Task {task_num} completed: {len(result.text)} chars")

        duration = time.time() - start
        print(f"\n✓ Parallel processing took: {duration:.2f}s")

        # Check queue stats
        if session.has_queue():
            stats = session.queue_stats()
            print("\n📊 Queue Statistics:")
            print(f"  Total processed: {stats['total_processed']}")
            print(f"  Total failed: {stats['total_failed']}")
            print(f"  DLQ size: {stats['dlq_size']}")

        print("\n✅ Test 2 passed: Parallel processing with queue works\n")

    async def test_priority_lanes(self) -> None:
        """Test 3: Parallel processing with priority lanes."""
        print("🎯 Test 3: Parallel Processing with Priority Lanes")
        print("-" * 80)

        queue_config = SessionQueueConfig()
        queue_config.set_control_concurrency(1)    # P0: Control operations
        queue_config.set_query_concurrency(3)      # P1: Query operations (highest concurrency)
        queue_config.set_execute_concurrency(2)    # P2: Execute operations
        queue_config.set_generate_concurrency(1)   # P3: Generate operations
        queue_config.enable_metrics()
        queue_config.enable_dlq()

        session = self.agent.session(".", queue_config=queue_config)

        print("Testing priority-based task execution...")
        print("  P1 (Query): 3 concurrent tasks")
        print("  P2 (Execute): 2 concurrent tasks")
        print("  P3 (Generate): 1 concurrent task")

        start = time.time()

        # Mix of different task types
        tasks = [
            ("Query", "How many Python files are there?"),
            ("Query", "What is the project structure?"),
            ("Query", "List all dependencies"),
            ("Execute", "Create a summary of the codebase"),
            ("Execute", "Analyze code complexity"),
        ]

        async def process_task(task_num: int, task_type: str, task_text: str) -> tuple:
            """Process a single task."""
            result = session.send(task_text)
            return task_num, task_type, result

        for i, (task_type, task) in enumerate(tasks, 1):
            print(f"  Queuing {task_type} task {i}: {task}")

        print("\n  Waiting for all tasks to complete...")

        async_tasks = [
            asyncio.create_task(asyncio.to_thread(process_task, i, task_type, task))
            for i, (task_type, task) in enumerate(tasks, 1)
        ]

        results = await asyncio.gather(*async_tasks, return_exceptions=True)

        for task_num, task_type, result in results:
            if isinstance(result, Exception):
                print(f"  ✗ Task {task_num} ({task_type}) failed: {result}")
            else:
                print(f"  ✓ Task {task_num} ({task_type}) completed: {len(result.text)} chars")

        duration = time.time() - start
        print(f"\n✓ Priority-based processing took: {duration:.2f}s")
        print("\n✅ Test 3 passed: Priority lanes work correctly\n")

    async def test_retry_policy(self) -> None:
        """Test 4: Parallel processing with retry policy."""
        print("🔄 Test 4: Parallel Processing with Retry Policy")
        print("-" * 80)

        queue_config = SessionQueueConfig()
        queue_config.set_query_concurrency(3)
        queue_config.enable_metrics()
        queue_config.enable_dlq()

        session = self.agent.session(".", queue_config=queue_config)

        print("Testing retry policy with exponential backoff...")
        print("  Note: Retry policy configured at queue level")
        print("  Max retries: 3 (default)")
        print("  Strategy: exponential (default)")

        tasks = [
            "Analyze the main function",
            "Find all error types",
            "List all test functions",
        ]

        start = time.time()

        async def process_task(task_num: int, task_text: str) -> tuple:
            """Process a single task."""
            result = session.send(task_text)
            return task_num, result

        for i, task in enumerate(tasks, 1):
            print(f"  Queuing task {i}: {task}")

        print("\n  Waiting for all tasks to complete...")

        async_tasks = [
            asyncio.create_task(asyncio.to_thread(process_task, i, task))
            for i, task in enumerate(tasks, 1)
        ]

        results = await asyncio.gather(*async_tasks, return_exceptions=True)

        for task_num, result in results:
            if isinstance(result, Exception):
                print(f"  ✗ Task {task_num} failed: {result}")
            else:
                print(f"  ✓ Task {task_num} completed: {len(result.text)} chars")

        duration = time.time() - start
        print(f"\n✓ Processing with retry took: {duration:.2f}s")

        if session.has_queue():
            stats = session.queue_stats()
            print("\n📊 Final Queue Statistics:")
            print(f"  Total processed: {stats['total_processed']}")
            print(f"  Total failed: {stats['total_failed']}")
            print(f"  DLQ size: {stats['dlq_size']}")

        print("\n✅ Test 4 passed: Retry policy works correctly\n")

    async def run_all(self) -> None:
        """Run all parallel processing tests."""
        print("🚀 A3S Code Python SDK - Parallel Task Processing Integration Test\n")
        print("=" * 80)

        print(f"📄 Using config: {self.config_path}")
        print("=" * 80)

        await self.test_sequential_processing()
        await self.test_parallel_processing()
        await self.test_priority_lanes()
        await self.test_retry_policy()

        print()
        print("=" * 80)
        print("✅ All parallel processing tests completed successfully!")
        print("=" * 80)


if __name__ == "__main__":
    config_path = ParallelProcessingTest.find_config_path()
    agent = Agent.create(config_path)
    test = ParallelProcessingTest(agent, config_path)
    asyncio.run(test.run_all())
